rq/worker.py

```diff
@@ -667,24 +669,36 @@ def perform_job(self, job, queue):
                 # to use the same exc handling when pickling fails
                 job._result = rv
 
-                self.set_current_job_id(None, pipeline=pipeline)
-
                 result_ttl = job.get_result_ttl(self.default_result_ttl)
                 if result_ttl != 0:
                     job.ended_at = utcnow()
-                    job.set_status(JobStatus.FINISHED, pipeline=pipeline)
-                    job.save(pipeline=pipeline)
 
-                    finished_job_registry = FinishedJobRegistry(job.origin,
-                                                                self.connection)
-                    finished_job_registry.add(job, result_ttl, pipeline)
+                while True:
+                    try:
+                        # if dependencies are inserted after enqueue_dependents
+                        # a WatchError is thrown by execute()
+                        pipeline.watch(job.dependents_key)
```

Comment:
The problem with the code above is that the protected code starts with redis' `WATCH` call. I don't know the used redis-library, so I don't know what happens to the statements before the `watch()` call, but it is probably wrong or undefined. It is up to the library what happens with the statements before the `watch()` call. So if a concurrent modification occurs, the library may even commit the changes before `watch()`, so the job in redis is actually FINISHED, but the loop would restart the transaction.
So the only way is to:

1. call the `watch()` command, which sets the connection into the immediate-execution mode
2. get all the necessary information: for us it is the dependency list together with the necessary metadata (happens in `enqueue_dependents`)
3. call `multi()` to get back into the pipelined mode for efficiency
4. enqueue all dependencies and finish the job
5. commit the transaction with `execute`

If the transaction fails, nothing is written to redis and we have to restart again. But this "all or nothing" principle protects us from any inconsistency.

Ok, I will move the code.